Add to_base64url Presto function

czentgr opened the pull request:
Resolves: https://github.com/facebookincubator/velox/issues/4330

czentgr commented:
Not sure what the problem with the Pyvelox build for Mac is. It built folly ok but then just traps on an assert:

```
  real	6m28.672s
  user	13m40.033s
  sys	1m14.148s
  + Finished running install_folly
  + echo 'All deps for Velox installed! Now try "make"'
  All deps for Velox installed! Now try "make"
  + echo 'To add cmake-format bin to your $PATH, consider adding this to your ~/.profile:'
  To add cmake-format bin to your $PATH, consider adding this to your ~/.profile:
  + echo 'export PATH=$HOME/bin:$HOME/Library/Python/3.7/bin:$PATH'
  export PATH=$HOME/bin:$HOME/Library/Python/3.7/bin:$PATH
  Traceback (most recent call last):
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/bin/cibuildwheel", line 8, in <module>
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/__main__.py", line 141, in main
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/__main__.py", line 297, in build_in_directory
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/macos.py", line 333, in build
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/options.py", line 531, in build_options
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/util.py", line 347, in with_defaults
    File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/util.py", line 342, in __init__
  AssertionError
  Error: Process completed with exit code 1.
```

aditi-pandit commented:
> Not sure what the problem with the Pyvelox build for Mac is. It built folly ok but then just traps on an assert:
> 
> ```
>   real	6m28.672s
>   user	13m40.033s
>   sys	1m14.148s
>   + Finished running install_folly
>   + echo 'All deps for Velox installed! Now try "make"'
>   All deps for Velox installed! Now try "make"
>   + echo 'To add cmake-format bin to your $PATH, consider adding this to your ~/.profile:'
>   To add cmake-format bin to your $PATH, consider adding this to your ~/.profile:
>   + echo 'export PATH=$HOME/bin:$HOME/Library/Python/3.7/bin:$PATH'
>   export PATH=$HOME/bin:$HOME/Library/Python/3.7/bin:$PATH
>   Traceback (most recent call last):
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/bin/cibuildwheel", line 8, in <module>
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/__main__.py", line 141, in main
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/__main__.py", line 297, in build_in_directory
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/macos.py", line 333, in build
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/options.py", line 531, in build_options
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/util.py", line 347, in with_defaults
>     File "/usr/local/Cellar/pipx/1.1.0/.cache/8c372460604e327/lib/python3.11/site-packages/cibuildwheel/util.py", line 342, in __init__
>   AssertionError
>   Error: Process completed with exit code 1.
> ```
> 
> Not clear to me what assert failed and how (if at all) it is related to my code.

Please rebase after https://github.com/facebookincubator/velox/pull/4366 and try.

aditi-pandit reviewed (approved):
Looks good. Thanks !

mbasmanova commented on velox/docs/functions/presto/binary.rst:
```diff
@@ -58,6 +58,10 @@ Binary Functions
 
+.. function:: to_base64url(binary) -> varchar
+
+    Encodes ``binary`` into a base64 string representation using the URL safe alphabet.
```
Perhaps, turn "URL safe alphabet" into a link to https://www.rfc-editor.org/rfc/rfc4648#section-5 to provide some context for folks who are not familiar with it.

mbasmanova commented on velox/functions/lib/string/StringImpl.h:
```diff
@@ -419,6 +419,15 @@ FOLLY_ALWAYS_INLINE bool fromBase64(
 }
 
+template <typename TOutString, typename TInString>
+FOLLY_ALWAYS_INLINE bool toBase64Url(
```
this method can be void

czentgr replied:
Right. I suppose the `fromBase64`/`toBase64` could be changed too. Not sure why they are declared with return type bool now that you point it out.

mbasmanova replied:
@czentgr Yes, it would be nice to change other functions as well (preferably, in a separate PR). We didn't always support void call functions. These may have been added before we added such support.

czentgr replied:
Sounds good, I can manage with a cleanup PR when also fixing the naming.

mbasmanova commented on velox/functions/prestosql/StringFunctions.h:
```diff
@@ -259,6 +259,17 @@ struct FromBase64Function {
+struct ToBase64UrlFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  FOLLY_ALWAYS_INLINE bool call(
```
It would be more efficient to make this method return 'void'.

mbasmanova commented on velox/common/encode/Base64.h:
```diff
@@ -71,6 +71,11 @@ class Base64 {
+  /// Encodes the specified number of characters from the 'data' and writes the
+  /// result to the 'output'. The output must have enough space, e.g. as
+  /// returned by the calculateEncodedSize().
+  static void encode_url(const char* data, size_t size, char* output);
```
naming: encode_url -> encodeUrl

I see that there are existing functions named encode_url. This was probably copy-pasted from somewhere. Let's add a TODO to rename these as well. Consider sending a separate PR for rename.

czentgr replied:
Added a TODO. Will do this in the next PR once this got merged.

mbasmanova reviewed (approved):
@czentgr Looks great % a couple small comments.

czentgr commented:
Fuzzer results after addressing the review comments:
```
I0322 13:16:05.599658 227249 ExpressionVerifier.cpp:57] ===================
I0322 13:16:05.599673 227249 ExpressionVerifier.cpp:68] All results match.
I0322 13:16:05.599709 227249 ExpressionFuzzer.cpp:1155] ==============================> Done with iteration 12067
I0322 13:16:05.599751 227249 ExpressionFuzzer.cpp:972] ==============================> Top 1 by number of rows processed
I0322 13:16:05.599754 227249 ExpressionFuzzer.cpp:974] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0322 13:16:05.599757 227249 ExpressionFuzzer.cpp:978] to_base64url 12068 100.00% 197095
I0322 13:16:05.599784 227249 ExpressionFuzzer.cpp:984] ==============================> Bottom 1 by number of rows processed
I0322 13:16:05.599786 227249 ExpressionFuzzer.cpp:986] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0322 13:16:05.599789 227249 ExpressionFuzzer.cpp:991] to_base64url 12068 100.00% 197095
I0322 13:16:05.599793 227249 ExpressionFuzzer.cpp:1004] ==============================> All stats sorted by number of times the function was chosen
I0322 13:16:05.599795 227249 ExpressionFuzzer.cpp:1006] Format: functionName numTimesSelected proportionOfTimesSelected numProcessedRows
I0322 13:16:05.599798 227249 ExpressionFuzzer.cpp:1010] to_base64url 12068 100.00% 197095
```

mbasmanova commented:
@czentgr Looks great to me. Thank you for addressing the comments, running the Fuzzer and sharing the results.
